File: app/src/main/java/com/zerofate/template/view/recyclerview/RecyclerViewTestActivity.java
package com.zerofate.template.view.recyclerview;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.zerofate.andoroid.data.Shakespeare;
import com.zerofate.template.R;
import com.zerofate.template.databinding.ActivityRecyclerViewTestBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import butterknife.ButterKnife;

public class RecyclerViewTestActivity extends AppCompatActivity {

    private static final String TAG = "RecyclerViewTestActivit";
    private ActivityRecyclerViewTestBinding binding;
    private List<String> datas;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_recycler_view_test);
        ButterKnife.bind(this);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        datas = getRandomStrings();
        binding.recyclerView.setAdapter(new RecyclerView.Adapter<CustomViewHolder>() {
            @NonNull
            @Override
            public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                    int viewType) {
                return new CustomViewHolder(LayoutInflater.from(RecyclerViewTestActivity.this).inflate(R.layout.item_text,null));
            }

            @Override
            public void onBindViewHolder(@NonNull CustomViewHolder holder,
                    int position) {
                Log.d(TAG, "onBindViewHolder: " + position);
                holder.textView.setText(datas.get(position));
            }

            @Override
            public int getItemCount() {
                return datas.size();
            }
        });
        binding.btnChangeData.setOnClickListener(v -> {
            List<String> newDatas = getRandomStrings();
            datas = newDatas;
            Log.d(TAG, "notify before ");
            binding.recyclerView.getAdapter().notifyDataSetChanged();
            Log.d(TAG, "onCreate: notify after");
        });

    }

    @NonNull
    private List<String> getRandomStrings() {
        List<String> newDatas = new ArrayList<>(5);
        for (int i = 0;i<5;i++){
            newDatas.add(Shakespeare.TITLES[new Random().nextInt(Shakespeare.TITLES.length)]);
        }
        return newDatas;
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "onDestroy: ");
        super.onDestroy();
    }

    private static class CustomViewHolder extends RecyclerView.ViewHolder{

        protected TextView textView;
        public CustomViewHolder(View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.text);
        }
    }


}

File: app/src/main/java/com/zerofate/template/view/recyclerview/RecyclerViewTestActivity.kt
package com.zerofate.template.view.recyclerview

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DividerItemDecoration
import com.dozeboy.android.core.utli.log.ZLog
import com.zerofate.android.data.Shakespeare
import com.zerofate.androidsdk.base.BaseSingleFragmentActivity
import com.zerofate.template.R
import kotlinx.android.synthetic.main.activity_recycler_view_test.*
import java.util.*

class RecyclerViewTestActivity : BaseSingleFragmentActivity() {


    override fun getFragment(startIntent: Intent?): androidx.fragment.app.Fragment {
        return RecyclerViewFragment()
    }

    override fun onDestroy() {
        Log.d(TAG, "onDestroy: ")
        super.onDestroy()
    }

    private class CustomViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        var textView: TextView = itemView.findViewById(R.id.text)
    }

    class RecyclerViewFragment : androidx.fragment.app.Fragment() {
        private var datas: List<String>? = null
        private val randomStrings: List<String>
            get() {
                val newDatas = ArrayList<String>(5)
                for (i in 0..100) {
                    newDatas.add(Shakespeare.TITLES[Random().nextInt(Shakespeare.TITLES.size)])
                }
                return newDatas
            }

        override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
        ): View? {
            ZLog.d("onCreateView: ")
            return inflater.inflate(R.layout.activity_recycler_view_test, container, false)
        }

        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            recycler_view.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(activity)
            datas = randomStrings
            val itemDecoration = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
            itemDecoration.setDrawable(context?.resources?.getDrawable(R.drawable.abc_list_divider_material)!!)
            recycler_view.addItemDecoration(itemDecoration)
            recycler_view.adapter = object : androidx.recyclerview.widget.RecyclerView.Adapter<CustomViewHolder>() {
                @SuppressLint("InflateParams")
                override fun onCreateViewHolder(
                    parent: ViewGroup,
                    viewType: Int
                ): CustomViewHolder {
                    return CustomViewHolder(
                        LayoutInflater.from(activity).inflate(
                            R.layout.item_text,
                            null
                        )
                    )
                }

                override fun onBindViewHolder(
                    holder: CustomViewHolder,
                    position: Int
                ) {
                    Log.d(TAG, "onBindViewHolder: $position")
                    holder.textView.text = resources.getString(R.string.network_api)
                }

                override fun getItemCount(): Int {
                    return datas!!.size
                }
            }

            btn_change_data.setOnClickListener {
                datas = randomStrings
                (recycler_view.adapter as RecyclerView.Adapter<*>).notifyDataSetChanged()
            }
        }

        override fun onDestroyView() {
            super.onDestroyView()
            ZLog.d("onDestroyView: ")
        }

        override fun onDestroy() {
            super.onDestroy()
            ZLog.d("onDestroy: ")
        }

        override fun onDetach() {
            super.onDetach()
            ZLog.d("onDetach: ")
        }
    }

    companion object {
        private const val TAG = "RecyclerViewTestActivit"
    }


}
